Separate endpoints for recv communicators from same source endpoint

rauteric opened the pull request:
This patch changes the plugin's endpoint creation behavior as follows:

1. All send communicators will use the same endpoint.
2. When creating an endpoint for receive communicators, first check if
any existing endpoint is not already connected to the source endpoint.
Only create a new recv endpoint if all existing endpoints are already
connected to the source endpoint.

Also has some supporting refactoring commits.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

AmedeoSapio commented:
bot:aws:retest

rauteric commented on src/nccl_ofi_rdma.c:
```diff
@@ -5204,6 +5204,15 @@ static int connect(nccl_net_ofi_ep_t *base_ep,
 
 static void ep_rail_release(nccl_net_ofi_ep_rail_t *rail, int dev_id)
 {
+	if (ofi_nccl_endpoint_per_communicator() != 0) {
```
Something seems wrong here. We always create the device-cq, but we only avoid freeing it in `ep_rail_release` if `ofi_nccl_endpoint_per_communicator()`. In that case, the `endpoint_per_comm == 0` case is probably broken: we can free the device-cq here and then not allocate it again.

Instead, we should probably only allocate the device-shared CQ if `endpoint_per_comm != 0`.

rauteric replied:
Actually, for simplicity, I think we should just always allocate the device-shared CQ, and never free it in `ep_rail_release`. We can free it in the device's finalize method, which will never be called by NCCL today.

rauteric replied:
Fixed in next rev

bwbarrett replied:
I'm fine with your second update.  I do object to allocating "device resources", which includes CQs, unless we're going to use them.

rauteric commented on tests/unit/ep_addr_list.c:
```diff
@@ -0,0 +1,79 @@
+	/* TODO these have no return value so just make sure it doesn't crash */
+	nccl_ofi_delete_ep_for_addr(ep_addr_list, (nccl_net_ofi_ep_t *)1);
+	nccl_ofi_delete_ep_for_addr(ep_addr_list, (nccl_net_ofi_ep_t *)2);
+	nccl_ofi_delete_ep_for_addr(ep_addr_list, (nccl_net_ofi_ep_t *)3); // (Doesn't exist)
```
Have a case where we try to retrieve an ep for an address after these deletions -- should return NULL.

rauteric commented on src/nccl_ofi_rdma.c:
```diff
@@ -5669,6 +5669,52 @@ static void release_device_ofi_resources(nccl_net_ofi_rdma_device_t *device)
+		return -EINVAL;
+	}
+
+	vf_idx = (int)strtol(ptr + 1, &endptr, 10);
```
This needs a fix to account for the new EFA GUID format.

rauteric commented on src/nccl_ofi_rdma.c:
```diff
@@ -5790,6 +5829,12 @@ nccl_net_ofi_rdma_device_create(nccl_net_ofi_plugin_t *plugin,
 	}
 
 	device->num_comm_ids = (uint32_t)NCCL_OFI_RDMA_MAX_COMMS;
+	nccl_ofi_init_ep_addr_list(&device->ep_addr_list);
```
Only if `endpoint_per_comm != 0`.

rauteric commented on src/nccl_ofi_rdma.c:
```diff
@@ -5379,6 +5410,10 @@ static int release_ep(nccl_net_ofi_ep_t *base_ep)
 	 */
 	if (ep->ref_cnt == 0) {
+
+		if (ofi_nccl_endpoint_per_communicator() != 0)
+			nccl_ofi_delete_ep_for_addr(device->ep_addr_list, &ep->base);
```
For non-thread-local-eps, also free the endpoint, since the comment above about the thread-local pointer preventing freeing doesn't apply.

rauteric commented on include/nccl_ofi_ep_addr_list.h:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Initialize an endpoint-address-set pair list
+ *
+ * NOTE: there is no destroy function because this list is stored with device,
+ * which has no destructor.
```
Do some cleanup in the new device finalize API (even if it won't be used by NCCL).

rauteric commented on src/nccl_ofi_rdma.c:
```diff
@@ -5204,6 +5204,15 @@ static int connect(nccl_net_ofi_ep_t *base_ep,
+		(instead of a cq per endpoint), set the rail->cq pointer to NULL
+		here so	that the cq isn't actually released in ep_release().
+		The cq will be released when the domain is cleaned up
+		(which it currently never is, because there's no
+		plugin fini() function */
```
Well now there is. ([d4a0fea](https://github.com/aws/aws-ofi-nccl/commit/d4a0feadc3f3760c4da45e441b0db66db62df995)) 😄 Use this.

rauteric commented on include/nccl_ofi_param.h:
```diff
@@ -228,6 +228,13 @@ OFI_NCCL_PARAM_INT(eager_max_size, "EAGER_MAX_SIZE", 8192);
+ * communicators.  If non-0, create a Libfabric endpoint per
+ * communicator.
+ */
+OFI_NCCL_PARAM_INT(endpoint_per_communicator, "ENDPOINT_PER_COMM", 1);
```
This actually needs to be disabled by default. Without https://github.com/ofiwg/libfabric/pull/10045, which hasn't even been merged yet, let alone released, this code will have a perf penalty.

rauteric commented:
Force-push to 98bf2b1 is a rebase on master.

rauteric commented:
Force-push to 8850320 address (my own) feedback

rauteric commented on src/nccl_ofi_rdma.c:
```diff
@@ -5762,19 +5810,29 @@ static nccl_net_ofi_rdma_device_rail_t *create_device_rail_array(struct fi_info
 			goto error;
 		}
 
+		vf_idx = get_rail_vf_idx(info_list);
+		if (vf_idx == -1) {
+			rail_idx = i;
+		} else {
+			rail_idx = rail_map[vf_idx];
+			rail_map[vf_idx]++;
+		}
```
I think we should restrict this change to only the EFA provider. (Or, just allow it to fail on other providers, but I'm not ecstatic about that option.)

bwbarrett replied:
Why?

rauteric replied:
It relies on a GUID available in `/sys/class/infiniband/%s/node_guid` as well as assumes a particular GUID format to locate vf_idx (https://github.com/aws/aws-ofi-nccl/pull/438/commits/3d1d05f25273fb04648ce71e0a6791d5351a83a7#diff-70b13370752f15bd8052cf01a6f9748fdd022b830a4f0aeacc10add288008540R5785-R5789), both of which I thought are EFA-specific.

bwbarrett replied:
and that's what I get for trying to look on my phone.  You're right, this needs to be in platform-aws.

rauteric replied:
Done

rauteric commented:
(Note: the CI failure is spurious. Will try again on next revision.)

rauteric commented:
Somehow I majorly butchered performance in yesterday's update to this PR. Need to sort that out.

rauteric commented:
Moved rail index assignment logic to platform hook and added AWS/EFA implementation.

rauteric commented on src/nccl_ofi_rdma.c:
```diff
@@ -1515,6 +1515,22 @@ static inline int process_err_completion(nccl_net_ofi_rdma_device_t *device,
+		req = err_entry.op_context;
+		if (req == NULL || req->type != NCCL_OFI_RDMA_BOUNCE) {
+			NCCL_OFI_WARN("Received canceled event from unexpected request type. Type: %s",
+				nccl_net_ofi_req_str(req));
```
Nit: can't call this function if `req == NULL`.

rauteric replied:
Fixed

bwbarrett commented on include/nccl_ofi_ep_addr_list.h:
```diff
@@ -0,0 +1,55 @@
+typedef struct nccl_ofi_ep_addr_list nccl_ofi_ep_addr_list_t;
+
+/**
+ * Initialize an endpoint-address-set pair list
```
who allocates memory?

rauteric replied:
This function does. Will note in docstring.

bwbarrett commented on tests/unit/ep_addr_list.c:
```diff
@@ -0,0 +1,107 @@
+		exit(1);
+	}
+
+	/** Now try retrieving for a new address -- should return NULL since we removed all eps **/
```
You kind of have an all or nothing approach.  you don't test cases like I've inserted a, b, and c, then remove b, but want to make sure a and c are still there.  Or any of the error cases.  We don't have code coverage metrics, but I'd bet this is testing less than 50% of the code you wrote.

rauteric replied:
I do agree that the remove-middle test is useful, I will add that.

Not sure about the code coverage bit -- from what I can tell, we're testing every path in nccl_ofi_ep_addr_list.c, except for the failures to allocate memory, which I don't have a good way to test.

bwbarrett commented on include/nccl_ofi.h:
```diff
@@ -265,9 +265,16 @@ struct nccl_net_ofi_device {
 	int (*get_ep)(nccl_net_ofi_device_t *base_dev,
-			       nccl_net_ofi_ep_t **ep);
+			       nccl_net_ofi_ep_t **ep,
+			       bool use_thread_local_ep);
```
isn't this a property of the device?  Why should it be an argument here?

rauteric replied:
Before this PR, `get_ep` will always return the thread-local ep. In this PR, I need a way to specify that I want to allocate a new EP and not use the thread-local EP.

Let me think of a cleaner API for this.

rauteric replied:
Cleaned this up by creating a separate function.

bwbarrett commented on include/nccl_ofi_rdma.h:
```diff
@@ -602,6 +603,9 @@ struct nccl_net_ofi_rdma_ep {
 	size_t bounce_buff_size;
+
+	/* True if this ep is stored in the thread-local store */
+	bool thread_local_ep;
```
i'm not sure I understand both why this is a property of the ep, nor why this is the right thing to test for.

rauteric replied:
This is a bit thorny. When calling `release_ep`, we can only call `free` on the endpoint reference if it is not stored in the device's thread-local ep store ([explanation](https://github.com/aws/aws-ofi-nccl/blob/aa4d36e549f12a9ca32f4ffb53cbe1e0e74b4e3f/src/nccl_ofi_rdma.c#L5367-L5382)). So, during `release_ep`, we need a way to know whether this is a thread-local endpoint, hence this property of the ep.

I couldn't immediately think of a better way.

bwbarrett replied:
Is there a different API that's wrong such that we end up in this corner of hell?

rauteric replied:
I suppose the ugliness is because `get_ep` and `release_ep` have the unexpected side effect of tinkering with a thread-local endpoint reference. We still need that behavior for the send communicators. Maybe we can revisit how we deal with the thread-local ep reference after this change.

bwbarrett replied:
You're suggesting that we add more bad code rather than fix previously created abstraction issues?

rauteric replied:
On abstraction issues, I think we have a couple of problems:
1. Currently (in master today), the API (nccl_ofi_api.c) "owns" the endpoints, and is responsible for creating and destroying the endpoints that are used by the communicators.
2. This PR is attempting to make recv_comm own its endpoint that it uses for communication (RDMA protocol only), but trying to keep compatibility with the API's semantics on endpoint creation and deletion, which needs this special case.

I think fixing this properly would take some refactoring effort to move endpoint ownership into the communicators instead of the API.

One simple thing we can do is set the `release_ep` member of endpoint appropriately, based on whether or not it is a thread-local ep. I.e., we'd have two release_ep methods: `release_ep` and `release_ep_thread_local`, and the `base_ep->release_ep` member function would be set to one of these.

bwbarrett replied:
When you're designing in the future, you need to take this kind of issue into account; the right answer is to do the required refactoring.  But for now, we'll deal with this.  I really hate the multiple release_ep ideas.

bwbarrett commented on src/nccl_ofi_api.c:
```diff
@@ -234,7 +234,7 @@ ncclResult_t nccl_net_ofi_connect(int dev_id, void *handle, void **sComm)
 		}
 
-		int ret = device->get_ep(device, &base_ep);
+		int ret = device->get_ep(device, &base_ep, true);
```
You should not be blindly passing true here; this is a sign you have the API wrong.

bwbarrett commented on src/nccl_ofi_rdma.c:
```diff
@@ -2126,6 +2127,15 @@ static int init_send_comm_rails(nccl_net_ofi_rdma_send_comm_t *s_comm,
 	}
 
+	/**
+	 * In ENDPOINT_PER_COMM config, the ep address in the handle is not
```
Put a TODO comment here.  We shouldn't be doing this, but it's not worth fixing now, given that we're going to do the control channel qp work soonish.

bwbarrett commented on src/nccl_ofi_rdma.c:
```diff
@@ -3794,6 +3820,22 @@ static int accept(nccl_net_ofi_listen_comm_t *listen_comm,
 		l_comm->r_comm = r_comm;
 
+		/* Just in case ep changed, reset it here */
+		ep = (nccl_net_ofi_rdma_ep_t *)r_comm->base.base.ep;
```
?

rauteric replied:
The handling of this ep reference was messy. Next revision attempts to make it better.

bwbarrett commented on include/Makefile.am:
```diff
@@ -50,4 +50,5 @@ noinst_HEADERS = \
 	nccl-headers/neuron/error.h \
 	contrib/uthash.h \
 	contrib/utlist.h \
-	nccl_ofi_ep_addr_list.h
```
In the "Move platform-specific hooks to nccl_ofi_platform.h" commit message, you need a body that explains why.

rauteric replied:
I didn't have any better than having a clean separation for these hooks... but added to commit message.

bwbarrett commented on include/nccl_ofi_platform.h:
```diff
@@ -27,4 +27,14 @@ int platform_init(const char **provider_filter) __attribute__((weak));
+ * @param num_rails: number of rails
+ * @param rail_indices: array of `num_rails` indices, set by this function
+ */
+void platform_get_rail_indices(struct fi_info *info_list, int num_rails,
```
This seems rather specific to us; are we sure there's not a more general interface here?

rauteric replied:
I can't think of one. Maybe `get_unique_guid(fi_info)`? Though that's not much different than what this function does.

bwbarrett replied:
what about platform_sort_rails()?

rauteric replied:
That would work, though I'm not sure it's meaningfully different from `platform_get_rail_indices`.

Moving the rail-sorting part to a separate PR.

rauteric commented on src/nccl_ofi_rdma.c:
```diff
@@ -1515,6 +1519,22 @@ static inline int process_err_completion(nccl_net_ofi_rdma_device_t *device,
+			ret = -EIO;
+			goto exit;
+		}
+		NCCL_OFI_TRACE(NCCL_NET, "Processed canceled event for bounce request %p", req);
```
Changed this print. I wanted to avoid the legwork of renaming all of these during this PR.

rauteric commented on src/nccl_ofi_rdma.c:
```diff
@@ -3736,6 +3762,7 @@ static int accept(nccl_net_ofi_listen_comm_t *listen_comm,
 		/* Progress NCCL OFI engine so that connection is accepted */
 		ret = ofi_process_cq(ep);
 		if (OFI_UNLIKELY(ret != 0)) {
+			NCCL_OFI_WARN("Failure to process cq: RC %d", ret);
```
I added this print in frustration after my run was failing without any error message. It turned out to be that I forgot to put a print at the point the error occurred somewhere inside `ofi_process_cq`.

But I would argue that these types of prints are useful for backtracking what failed in cases we forget to put a print where the original error happens.

rauteric replied:
But in any case, as you mention in your next comment, these prints don't belong in this commit. So will remove.

rauteric commented on include/nccl_ofi_ep_addr_list.h:
```diff
@@ -0,0 +1,55 @@
+#ifndef NCCL_OFI_EP_ADDR_LIST_H
+#define NCCL_OFI_EP_ADDR_LIST_H
+
+#include "nccl_ofi.h"
```
For `nccl_net_ofi_ep_t`.

I made the ep_addr_list take in pointers of this type for the endpoint to have some semblance of type checking (I'm trying to minimize use of void pointers). But I'm not ecstatic about using `nccl_net_ofi_ep_t`, because it suggests this structure is doing something specific to endpoints, whereas in reality it's treating the endpoint as an opaque pointer.

Suggestions welcome.

bwbarrett replied:
if it's just an opaque pointer, use a void*.  But also, since you don't actually look at the endpoint, you can just declare `struct nccl_net_ofi_ep_t;' in this file and not bring in all of nccl_ofi.h.

rauteric replied:
I took your second suggestion.

rauteric commented on src/nccl_ofi_rdma.c:
```diff
@@ -6048,6 +6091,13 @@ int nccl_net_ofi_rdma_init(const char *provider_filter,
 	eager_max_size = (size_t) ofi_nccl_eager_max_size();
 
+	if (domain_per_thread == 1 && ofi_nccl_endpoint_per_communicator() != 0) {
+		/* TODO support this configuration */
```
Created.

rauteric commented:
Force-push to 73fab34 addressed feedback.

rauteric commented:
Force-push to 205a026 is rebase on master.

bwbarrett reviewed (changes requested):
Please move the rail ordering patches to their own PR.  They're only tangentially related to this PR.

I would reorder the commits so that you can squash the "Create an endpoint per communicator" and "rdma: use ep_addr_list for endpoint-per-comm code" commits.

rauteric commented:
* Removed the rail-sorting patches from this PR (will post a separate PR)
* Squashed the ep-per-comm and use-ep-addr-list commits
* Removed include of nccl_ofi.h from ep addr list code.

bwbarrett commented on include/nccl_ofi_ep_addr_list.h:
```diff
@@ -0,0 +1,64 @@
+ *
+ * @param list: output, pointer to the new list
+ */
+void nccl_ofi_init_ep_addr_list(nccl_ofi_ep_addr_list_t **list);
```
I didn't notice this before (sorry), but these should really be named `<prefix>_<object>_<action>`.  So this should really be `nccl_ofi_ep_addr_list_init()`.  And likewise for all the other functions.

bwbarrett replied:
Also, why have this as the function signature?  Either it should return an error code or return a pointer to the allocated thing, but this option just seems hard to use for no good reason.

rauteric replied:
Agreed, and agreed.

bwbarrett commented on src/nccl_ofi_ep_addr_list.c:
```diff
@@ -0,0 +1,181 @@
+	nccl_net_ofi_ep_t *ep;
+	/* Hash set of addresses the endpoint is connected to */
+	hashed_addr_t *addr_set;
+} pair_ep_addr_set_t;
```
what's the difference between a `pair_ep_addr_set_t` and a `ep_pair_list_elem_t`?  Why are the names formatted differently?  Why are there two structures at all?

rauteric replied:
Good point, these can be combined into the same structure. Will do.

bwbarrett commented on src/nccl_ofi_ep_addr_list.c:
```diff
@@ -0,0 +1,181 @@
+
+	DL_FOREACH(ep_list->ep_pair_list, ep_pair) {
+		hashed_addr_t *found_handle;
+		HASH_FIND(hh, ep_pair->pair.addr_set, (char *)addr, MAX_EP_ADDR, found_handle);
```
This doesn't make sense.  There's no fixed size of an address, and you definitely can't assume it's MAX_EP_ADDR, so you're really just hashing junk here.  I'm really surprised this works; you're getting lucky by some calloc somewhere.

rauteric replied:
This was intentional. We call `fi_getname` with output to `ep_rail`, which is a member of `ep`, which is calloc'd. So all our addresses are zero-padded up to `MAX_EP_ADDR`.

Passing around the lengths returned by `fi_getname` would be a lot of extra legwork. We'd have to store it in the rail info, the connect message, and with each address in ep_addr_list.

Alternatively, we could just manually zero-out the extra bits immediately after calling `fi_getname` to make it more obvious that the address is zero-padded?

bwbarrett replied:
do it right; copy around the length.

rauteric replied:
Done

rauteric commented on include/nccl_ofi_rdma.h:
```diff
@@ -361,6 +361,7 @@ typedef struct nccl_net_ofi_rdma_req {
  */
 typedef struct nccl_ofi_rdma_ep_name {
 	char ep_name[MAX_EP_ADDR];
+	size_t ep_name_len;
```
I like storing `ep_name_len` in the same struct as `ep_name`. But this will cause some extra padding since `MAX_EP_ADDR` is not a multiple of 8.

This is not fast-path code so I didn't feel strongly about it.

bwbarrett replied:
i wouldn't worry about it.

rauteric commented on src/nccl_ofi_rdma.c:
```diff
@@ -4767,6 +4826,8 @@ static void prepare_send_connect_message(nccl_net_ofi_rdma_ep_t *ep, int dev_id,
 		memcpy(conn_msg->ep_names[rail_id].ep_name,
 		       ep->rails[rail_id].local_ep_name,
 		       ep->rails[rail_id].local_ep_name_len);
+		conn_msg->ep_names[rail_id].ep_name_len =
+			ep->rails[rail_id].local_ep_name_len;
```
Grr. This ended up in the wrong git commit.

rauteric commented:
Rebased master

rauteric commented:
bot:aws:retest

rajachan commented on src/nccl_ofi_ep_addr_list.c:
```diff
@@ -0,0 +1,221 @@
+	if (!new_pair) {
+		NCCL_OFI_WARN("Failed to allocate new ep list element");
+		ret = -ENOMEM;
+		goto unlock;
```
free(new_addr)?

rajachan commented on src/nccl_ofi_rdma.c:
```diff
@@ -1502,6 +1506,21 @@ static inline int process_err_completion(nccl_net_ofi_rdma_device_t *device,
+		/* Closing an EP with posted receives can generate cancellation events
+		   for the posted receives */
+		req = err_entry.op_context;
+		if (req == NULL || req->type != NCCL_OFI_RDMA_BOUNCE) {
```
If we are expecting a cancelation completion and squashing it, we should just explicitly cancel it ourselves before closing the EP in my opinion, so any other error completion indicates a problem elsewhere. But this works too. Hard to tell from this commit's diff, but to make sure, we already free up state for these requests right?

rauteric replied:
> If we are expecting a cancelation completion and squashing it, we should just explicitly cancel it ourselves before closing the EP in my opinion, so any other error completion indicates a problem elsewhere.

Yeah, I thought about that originally and discussed with Brian. We figured that closing the endpoint basically does the same thing, and fetching and cancelling the individual requests would be a decent amount of extra work.

>  Hard to tell from this commit's diff, but to make sure, we already free up state for these requests right?

Yes. Thanks for pointing this out: I think this is actually a race condition. We've already freed the bounce buffer requests at this point ([link](https://github.com/rauteric/aws-ofi-nccl/blob/3291241d091848ff07b18b8de79815f8d43844b6/src/nccl_ofi_rdma.c#L5528)), so we shouldn't be accessing the request here.

Instead, here in the cancel handling code, we need to just check that the event is a receive op, since we only have one type of receive op in the plugin.

rauteric replied:
> Yeah, I thought about that originally and discussed with Brian. We figured that closing the endpoint basically does the same thing, and fetching and cancelling the individual requests would be a decent amount of extra work.

Actually, we decided to go this route and explicitly cancel each recv req.

bwbarrett commented on src/nccl_ofi_rdma.c:
```diff
@@ -5431,8 +5506,12 @@ static int release_ep(nccl_net_ofi_ep_t *base_ep)
-		/* Ideally we would "un-post" the bounce buffers, but this
-		   should be accomplished by closing the endpoint. */
+
+		ret = cancel_posted_recv_ops(ep);
```
you should have a comment here about why we're doing this.

bwbarrett replied:
Also, please confirm you've opened the two tickets we talked about with the Libfabric team?

rauteric replied:
> you should have a comment here about why we're doing this.
Done

> Also, please confirm you've opened the two tickets we talked about with the Libfabric team?
Confirmed.

bwbarrett commented on src/nccl_ofi_rdma.c:
```diff
@@ -5544,6 +5635,101 @@ static int release_ep(nccl_net_ofi_ep_t *base_ep)
+	return ep;
+}
+
+static int initialize_ep(nccl_net_ofi_rdma_ep_t *ep,
```
This does more than initialize an endpoint, since it's allocating memory for rails.  I guess i'm not sure why we'd do that here instead of part of alloc?

rauteric replied:
Agree, the distinction was confusing. Consolidated this into one function called `create_ep`.

Long term, I'll create a backlog item to refactor the ownership model of the endpoint. Then we can remove some of the ugliness around `ref_cnt`.

bwbarrett commented on src/nccl_ofi_rdma.c:
```diff
@@ -5544,6 +5635,101 @@ static int release_ep(nccl_net_ofi_ep_t *base_ep)
+	return ret;
+}
+
+static int create_new_ep(nccl_net_ofi_rdma_device_t *device,
```
How is this different than allocate_ep and why have both?

rauteric replied:
Consolidated into `create_ep`.

sunkuamzn commented:
bot:aws:retest

a-szegel commented:
bot:aws:retest
